community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexPopulator.java

```diff
@@ -0,0 +1,260 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.index.schema;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+
+import org.neo4j.collection.primitive.PrimitiveLongSet;
+import org.neo4j.index.internal.gbptree.GBPTree;
+import org.neo4j.index.internal.gbptree.Layout;
+import org.neo4j.index.internal.gbptree.RecoveryCleanupWorkCollector;
+import org.neo4j.index.internal.gbptree.Writer;
+import org.neo4j.io.pagecache.IOLimiter;
+import org.neo4j.io.pagecache.PageCache;
+import org.neo4j.kernel.api.exceptions.index.IndexEntryConflictException;
+import org.neo4j.kernel.api.index.IndexEntryUpdate;
+import org.neo4j.kernel.api.index.IndexPopulator;
+import org.neo4j.kernel.api.index.IndexUpdater;
+import org.neo4j.kernel.api.index.PropertyAccessor;
+import org.neo4j.kernel.impl.index.GBPTreeUtil;
+
+import static org.neo4j.index.internal.gbptree.GBPTree.NO_HEADER;
+import static org.neo4j.index.internal.gbptree.GBPTree.NO_MONITOR;
+import static org.neo4j.kernel.api.schema.OrderedPropertyValues.ofUndefined;
+
+/**
+ * {@link IndexPopulator} backed by a {@link GBPTree}.
+ *
+ * @param <KEY> type of {@link SchemaNumberKey}.
+ * @param <VALUE> type of {@link SchemaNumberValue}.
+ */
+public abstract class NativeSchemaIndexPopulator<KEY extends SchemaNumberKey, VALUE extends SchemaNumberValue>
+        implements IndexPopulator
+{
+    static final byte BYTE_ONLINE = 1;
+    static final byte BYTE_FAILED = 0;
+
+    private final PageCache pageCache;
+    private final File storeFile;
+    private final KEY treeKey;
+    private final VALUE treeValue;
+    private final RecoveryCleanupWorkCollector recoveryCleanupWorkCollector;
+    private final ConflictDetectingValueMerger<VALUE> conflictDetectingValueMerger;
+    protected final Layout<KEY,VALUE> layout;
+
+    private Writer<KEY,VALUE> singleWriter;
+    private boolean dropped;
+
+    GBPTree<KEY,VALUE> tree;
+
+    NativeSchemaIndexPopulator( PageCache pageCache, File storeFile, Layout<KEY,VALUE> layout,
+            RecoveryCleanupWorkCollector recoveryCleanupWorkCollector )
+    {
+        this.pageCache = pageCache;
+        this.storeFile = storeFile;
+        this.layout = layout;
+        this.treeKey = layout.newKey();
+        this.treeValue = layout.newValue();
+        this.recoveryCleanupWorkCollector = recoveryCleanupWorkCollector;
+        this.conflictDetectingValueMerger = new ConflictDetectingValueMerger<>();
+    }
+
+    @Override
+    public synchronized void create() throws IOException
```

Comment:
We got into scenarios around fields and volatility since the populator is managed by at least two threads, life cycle wise. It was a simplistic approach to synchronize all lifecycle methods, methods which are called once. Many other things gets simpler that way.